server/src/main/java/org/elasticsearch/indices/SystemIndices.java

```diff
@@ -269,5 +314,56 @@ public String getDescription() {
         public Collection<String> getAssociatedIndexPatterns() {
             return associatedIndexPatterns;
         }
+
+        public TriConsumer<ClusterService, Client, ActionListener<ResetFeatureStateStatus>> getCleanUpFunction() {
+            return cleanUpFunction;
+        }
+
+        /**
+         * Clean up the state of a feature
+         * @param indexDescriptors List of descriptors of a feature's system indices
+         * @param associatedIndexPatterns List of patterns of a feature's associated indices
+         * @param name Name of the feature, used in logging
+         * @param clusterService A clusterService, for retrieving cluster metadata
+         * @param client A client, for issuing delete requests
+         * @param listener A listener to return success or failure of cleanup
+         */
+        public static void cleanUpFeature(
+            Collection<SystemIndexDescriptor> indexDescriptors,
+            Collection<String> associatedIndexPatterns,
+            String name,
+            ClusterService clusterService,
+            Client client,
+            ActionListener<ResetFeatureStateStatus> listener) {
+            List<String> systemIndices = indexDescriptors.stream()
+                .map(sid -> sid.getMatchingIndices(clusterService.state().getMetadata()))
+                .flatMap(List::stream)
+                .collect(Collectors.toList());
+
+            List<String> associatedIndices = new ArrayList<>(associatedIndexPatterns);
+
+            List<String> allIndices = Stream.concat(systemIndices.stream(), associatedIndices.stream())
+                .collect(Collectors.toList());
```

Comment:
I made the suggested change, but you're probably right about associatedIndexPatterns. I'll take a look and report back.

Reply:
This works as written, because we can pass a valid index pattern to the delete request, which will then handle the resolution logic when executed. We can't do the same thing with system index descriptor patterns because they can use character-class regexes, for example `.test-[ab]*`.

It might be worth resolving the associated indices here so that we can skip submitting the delete request when no indices exist. Do you think this is important to do?

Reply:
I think it is fine as it is unless we discover issues